doc/reference.md

```diff
@@ -0,0 +1,509 @@
+
+## Expressions
+
+Jsonnet programs are composed entirely out of [expressions](https://en.wikipedia.org/wiki/Expression_(computer_science)). There are no statements or special top-level declarations present in most other languages. For instance, imports, conditionals, functions,  objects, and locals are all expressions.
+
+Any kind of expression can be a Jsonnet program, it is not necessary to have a top-level object. For example `2+2`, `"foo"` or `local bar = 21; bar * 2` are all valid Jsonnet programs.
+
+Expressions can be *evaluated* to produce a *value*. The evaluation does not have any side effects.
+
+The value to which an expression evaluates depends on its *environment* – values of variables, to which the expression refers. For example expression `x * x` depends on the value of `x`. Such expression is only valid in contexts, where variable `x` is available, e.g. `local x = 2; x * x` or `function(x); x * x`.
+
+Jsonnet variables follow the rules of *lexical scoping*. The environment mentioned before and consequently the meaning of variables in a given expression is statically determined:
+
+```
+local a = local x = 'a'; x;
+local foo = local x = 'b';  a;
+foo
+```
+
+A special case of that is creating [closures](https://en.wikipedia.org/wiki/Closure_(computer_programming)), such as the following:
+```
+local addNumber(number) = function(x) number + x;
+local add2 = addNumber(2);
+local add3 = addNumber(3);
+[
+  add2(2),
+  add3(5)
+]
+```
+
+## Values
+
+Jsonnet has only seven types of values:
+* `null` – only one value, namely `null`.
+* `boolean` – two values, `true` and `false`. 
+* `string` – Unicode strings (sequences of Unicode codepoints).
+* `number` – IEEE754 64-bit floating point number.
+* `function`, pure functions which take values as arguments and return a value.
+* `array`, a finite-length array of values. 
+* `object`, a superset of JSON objects with support for inheritance.
+
+All Jsonnet values are immutable. It is not possible to change a value of a field of an object or an element of an array – you can only create a new one with the desired change applied. 
+
+You can check the type of any value using `std.type`.
+
+### Equivalence and Equality
+
+We say that values `a` and `b` are equal if `a == b` evaluates to `true` and that they are unequal if `a == b` evaluates to `false`. Some pairs of values are neither equal or unequal, because functions cannot be checked for equality (and consequently some complex values containing functions). 
+
+Values of different types are never considered equal – there is no implicit casting (unlike, for example, in JavaScript).
+
+We say that values `a` and `b` are *the same value*, or *equivalent* if it is impossible to tell these values apart in any way in Jsonnet. More formally `a` and `b` are equivalent if there does not exist a Jsonnet function `f` such that exactly one of `f(a)` and `f(b)` results in an error.
```

Comment:
Yes, I think it's a very useful concept. 

This is just a more precise version intuitive notion that a value is "the same". It is formally, so that there is no question what counts as equivalent or not (e.g. it does not matter whether some array elements are already cached). 

The formal definition may look a bit weird, but it captures the boundary cases, such as:
`function(x) x * x` vs `function(x) if x == 42 then error 'err' else x * x`. 

We use it explicitly when talking about the properties of inheritance and in a few other places. We depend on it implicitly everywhere – even simple statements like `bool` has only two values `true` and `false` requires some notion of equivalence. Otherwise we wouldn't be able to say that  `2 == 2` and `true` are the same thing.